hare4/hare.go

```diff
@@ -283,6 +308,135 @@ func (h *Hare) Running() int {
 	return len(h.sessions)
 }
 
+// fetchFull will fetch the full list of proposal IDs from the provided peer.
+func (h *Hare) fetchFull(ctx context.Context, peer p2p.Peer, msgId types.Hash32) (
+	[]types.ProposalID, error,
+) {
+	ctx, cancel := context.WithTimeout(ctx, fetchFullTimeout)
+	defer cancel()
+
+	requestCompactCounter.Inc()
+	req := &CompactIdRequest{MsgId: msgId}
+	reqBytes := codec.MustEncode(req)
+	resp := &CompactIdResponse{}
+	cb := func(ctx context.Context, rw io.ReadWriter) error {
+		respLen, _, err := codec.DecodeLen(rw)
+		if err != nil {
+			return fmt.Errorf("decode length: %w", err)
+		}
+		if respLen >= MAX_EXCHANGE_SIZE {
+			return errResponseTooBig
+		}
+		buff := make([]byte, respLen)
+		_, err = io.ReadFull(rw, buff)
+		if err != nil {
+			return fmt.Errorf("read response buffer: %w", err)
+		}
+		err = codec.Decode(buff, resp)
+		if err != nil {
+			return fmt.Errorf("decode response: %w", err)
+		}
+		return nil
+	}
+
+	err := h.p2p.StreamRequest(ctx, peer, reqBytes, cb)
+	if err != nil {
+		requestCompactErrorCounter.Inc()
+		return nil, fmt.Errorf("stream request: %w", err)
+	}
+
+	h.tracer.OnCompactIdResponse(resp)
+
+	return resp.Ids, nil
+}
+
+func (h *Hare) handleProposalsStream(ctx context.Context, msg []byte, s io.ReadWriter) error {
+	requestCompactHandlerCounter.Inc()
+	compactProps := &CompactIdRequest{}
+	if err := codec.Decode(msg, compactProps); err != nil {
+		malformedError.Inc()
+		return fmt.Errorf("%w: decoding error %s", pubsub.ErrValidationReject, err.Error())
+	}
+	h.tracer.OnCompactIdRequest(compactProps)
+	h.mu.Lock()
+	m, ok := h.messageCache[compactProps.MsgId]
+	h.mu.Unlock()
+	if !ok {
+		messageCacheMiss.Inc()
+		return fmt.Errorf("message %s: cache miss", compactProps.MsgId)
+	}
+	resp := &CompactIdResponse{Ids: m.Body.Value.Proposals}
+	respBytes := codec.MustEncode(resp)
```

Comment:
As above, can't we just encode `resp` into the stream instead of using an intermediate buffer?

Reply:
To do this one must know to which length the type is going to encode into without really encoding it (since you must encode how many bytes are in the response before actually writing them into the response). While it is possible, it's probably something that the scale package should offer instead of writing it by hand which would be fragile. Any ideas on how to do this?

Reply:
Why do you need to encode the length first?

Reply:
Because you need to know how many bytes to read out of the stream.

Reply:
AFAIR, the scale decoding would fail if the response had the wrong length anyway.

Reply:
well, while we could do smth like this in `go-scale`, it is not really fragile if the data are an array of fixed-size IDs.
pls check how it's done in the `fetch` package, e.g. epoch stream handler
https://github.com/spacemeshos/go-spacemesh/blob/13dfb49679b475ac2e600e64872d9f0acd32dcc6/fetch/handler.go#L115
https://github.com/spacemeshos/go-spacemesh/blob/13dfb49679b475ac2e600e64872d9f0acd32dcc6/fetch/handler.go#L140-L185
and the client part
https://github.com/spacemeshos/go-spacemesh/blob/13dfb49679b475ac2e600e64872d9f0acd32dcc6/fetch/mesh_data.go#L275
https://github.com/spacemeshos/go-spacemesh/blob/13dfb49679b475ac2e600e64872d9f0acd32dcc6/fetch/mesh_data.go#L461-L481
The client part is not perfect b/c we could be handling each ID right away instead of waiting for the whole slice to be sent, but that would require further refactoring of the fetcher code; not sure whether it is applicable here.
Nevertheless, the server side can be improved here I think

Maybe we could move some helpers to the `codec` package

Reply:
> AFAIR, the scale decoding would fail if the response had the wrong length anyway.

So the reason it is done this way is the following:
When you read a response from a peer you want to avoid a situation where you read a stream until an EOF - this is generally bad because you put yourself in the risk of having a malicious peer to just feed you data which is read all into memory, resulting in the node memory usage going through the roof.
That's why you want to know how long is the data in advance.

Reply:
@ivan4th the `server.ReadResponse` part implicitly reads the `respLen` which encodes the whole message size in advance https://github.com/spacemeshos/go-spacemesh/blob/develop/p2p/server/server.go#L511